lib/yargs-factory.ts

```diff
@@ -387,6 +387,13 @@ export class YargsInstance {
         yargs: YargsInstance
       ): Partial<Arguments> | Promise<Partial<Arguments>> => {
         let aliases: Dictionary<string[]>;
+
+        // Skip coerce logic if related arg was not provided
+        const shouldCoerce = Object.prototype.hasOwnProperty.call(argv, keys);
```

Comment:
I believe this can just be:
```
Object.hasOwnProperty.call(obj, 'a')
```

Reply:
I'm afraid this is not correct and it should be `Object.prototype.hasOwnProperty`.

The changes relating to this should be reverted.

Reply:
`Object.hasOwnProperty.call({a: 5}, 'a')` and `Object.prototype.hasOwnProperty.call({a: 5}, 'a')` both return `true`. (I thought there might be an issue when using a null object, but both return `false` when `Object.create(null)` is used as the first argument.) I don't know of any scenarios where their behavior differs. Do you have an example where it could cause problems?

Reply:
Also, `Object.hasOwnProperty.call === Object.prototype.hasOwnProperty.call` is true. I'm assuming they are a reference to the same object

Reply:
Notwithstanding the fact that this convention of using the `prototype` instead of the `Object` class directly is widely adopted.

It's the same reason to not use `{a: 1}.hasOwnProperty('a')`. The function could have been overridden via a global polyfill or patch meaning you're not using native behaviour.

I assume that `{}.hasOwnProperty.call === Object.hasOwnProperty.call` as well as the prototype. The point is that they *can* be overridden and are, thus, unreliable to use unless you use the underlying prototype.

Reply:
I mean, `Object.prototype` can just as easily be overwritten:

```
> Object.prototype.hasOwnProperty = () => 'batman'
[Function (anonymous)]
> {}.hasOwnProperty()
'batman'
```

We would have to rewrite all of yargs to use primordials that are frozen at Node.js' startup to really avoid this issue.

Reply:
> We would have to rewrite all of yargs to use primordials that are frozen at Node.js' startup to really avoid this issue.

This is how Node.js itself approaches the problem:

https://github.com/nodejs/node/blob/master/lib/internal/per_context/primordials.js

But it's made easier for them by the fact that they can do so during bootstrapping.

Reply:
Yeah, this is definitely not an issue that is limited to `yargs`, for better or worse it's a language issue.

Whilst I think overriding the built-in methods like `Object.hasOwnProperty()` is pretty crazy, for those crazy enough to do it I think there is enough expectation that the prototype is left alone that it can be used reliably. Hence why it's enforced by standard linting rules in eslint and recommendations elsewhere.

Reply:
@dhensby makes sense to me.